### PR TITLE
AP_Scripting: allow scripting to get mission item/s by index, and wp data, and details of wp 

### PR DESCRIPTION
**Pull request by davidbuzz**
… from lua

### CODE REVIEW

**Comment by davidbuzz** [comment]
example lua snippet that loop over all mission items and sends them all to the user as sent_text messages, just because.   Warming, doing htis specific example with big mission/s will cause lua to timeout or exceed limits etc.

`
 missionsize = mission:num_commands()
  gcs:send_text(0, string.format("LUA: missionsize %d", missionsize ))    
  for x = 1, missionsize do
   gcs:send_text(0, string.format("LUA: mission-item %d", x ))    
    mitem = mission:get_waypoint_lua(x)
     gcs:send_text(0, string.format("LUA p1-4 %f,%f,%f,%f", mitem:param1(),mitem:param2(),mitem:param3(),mitem:param4() ))
     gcs:send_text(0, string.format("LUA x,y,z %0.6f,%0.6f,%0.2f", mitem:x()/10000000.0, mitem:y()/10000000.0, mitem:z() ))
     gcs:send_text(0, string.format("LUA seq,cmd,frm,cur %d,%d,%d,%d", mitem:seq(),mitem:command(),mitem:frame(),mitem:current() ))
 `
  end

File: libraries/AP_Mission/AP_Mission.cpp [hunk 1]
```diff
@@ -477,6 +477,73 @@ bool AP_Mission::set_current_cmd(uint16_t index)
     return true;
 }
 
+// returns false on any issue at all.
+bool AP_Mission::set_item(uint16_t index, mavlink_mission_item_int_t& src_packet)
+{
+        // this is the on-storage format
```
Comment (rmackay9) [inline libraries/AP_Mission/AP_Mission.cpp:483]: Looks like the indenting is 8spaces instead of 4?

File: libraries/AP_Mission/AP_Mission.cpp [hunk 2]
```diff
@@ -477,6 +477,73 @@ bool AP_Mission::set_current_cmd(uint16_t index)
...
+        AP_Mission::Mission_Command cmd;
+
+        // can't handle request for anything bigger than the mission size+1...
+        if (index > num_commands() ) {  return false; }
```
Comment (rmackay9) [inline libraries/AP_Mission/AP_Mission.cpp:487]: normally we would put the return false on a new line.  it's a small thing so not a blocker.

File: libraries/AP_Scripting/generator/description/bindings.desc [hunk 3]
```diff
@@ -205,3 +205,21 @@ singleton AP_Mission method get_prev_nav_cmd_id uint16_t
...
+userdata mavlink_mission_item_int_t field param1 float read write -FLT_MAX FLT_MAX
+userdata mavlink_mission_item_int_t field param2 float read write -FLT_MAX FLT_MAX
+userdata mavlink_mission_item_int_t field param3 float read write -FLT_MAX FLT_MAX
+userdata mavlink_mission_item_int_t field param4 float read write -FLT_MAX FLT_MAX
+userdata mavlink_mission_item_int_t field x int32_t read write -INT32_MAX INT32_MAX
+userdata mavlink_mission_item_int_t field y int32_t read write -INT32_MAX INT32_MAX
+userdata mavlink_mission_item_int_t field z float read write -FLT_MAX FLT_MAX
+userdata mavlink_mission_item_int_t field seq uint16_t read write 0 UINT16_MAX
+userdata mavlink_mission_item_int_t field command uint16_t read write 0 UINT16_MAX
+-- userdata mavlink_mission_item_int_t field target_system uint8_t read write 0 UINT8_MAX
```
Comment (rmackay9) [inline libraries/AP_Scripting/generator/description/bindings.desc:220]: I guess these are comment out lines?  normally we don't leave comments out lines in the code..

File: libraries/AP_Mission/AP_Mission.cpp [hunk 4]
```diff
@@ -477,6 +477,73 @@ bool AP_Mission::set_current_cmd(uint16_t index)
...
+        // setting ret_packet.command = -1  and/or returning false 
+        //  means it contains invalid data after it leaves here.
+
+        // this is the on-storage format
+        AP_Mission::Mission_Command cmd;
+
+        // can't handle request for anything bigger than the mission size...
+        if (index >= num_commands() ) { ret_packet.command = -1;  return false; }
+
+        mavlink_mission_request_int_t packet{}; // minimal placeholder
```
Comment (tridge) [inline libraries/AP_Mission/AP_Mission.cpp:513]: why is packet var needed at all?

Comment (davidbuzz) [reply libraries/AP_Mission/AP_Mission.cpp:513]: agreed, this var is no longer required, fixing it now.

File: libraries/AP_Mission/AP_Mission.cpp [hunk 5]
```diff
@@ -477,6 +477,73 @@ bool AP_Mission::set_current_cmd(uint16_t index)
...
+bool AP_Mission::get_item(uint16_t index, mavlink_mission_item_int_t& ret_packet)
+{
+        // setting ret_packet.command = -1  and/or returning false 
+        //  means it contains invalid data after it leaves here.
+
+        // this is the on-storage format
+        AP_Mission::Mission_Command cmd;
+
+        // can't handle request for anything bigger than the mission size...
+        if (index >= num_commands() ) { ret_packet.command = -1;  return false; }
```
[the remark this answers is not in the document]
Comment (davidbuzz) [reply libraries/AP_Mission/AP_Mission.cpp:511]: ran thru astyle, fixed.

File: libraries/AP_Mission/AP_Mission.cpp (hunk 2 above)
Comment (davidbuzz) [reply libraries/AP_Mission/AP_Mission.cpp:487]: ran thru astyle, fixed.

File: libraries/AP_Mission/AP_Mission.cpp (hunk 1 above)
Comment (davidbuzz) [reply libraries/AP_Mission/AP_Mission.cpp:483]: ran thru astyle, fixed.

File: libraries/AP_Scripting/generator/description/bindings.desc (hunk 3 above)
Comment (davidbuzz) [reply libraries/AP_Scripting/generator/description/bindings.desc:220]: in this particular instance they are here as a self-documenting reminder that one-day these other attributed might be useful, but right now they aren't used.

**Comment by davidbuzz** [comment]
> needs to run code style fixer (or manually)

done

**Comment by davidbuzz** [comment]
all previous requests have been fixed. and i rebased on master today to hopefully get current CI to pass.   .... hopefully.

**Comment by tridge** [comment]
needs approval from @WickedShell

**Comment by davidbuzz** [comment]
split bindings regeneration into two separate commits as per @tridge request.

**Comment by tridge** [comment]
merging after discussion on dev call

**Pull request merged**